sui_types/src/readable_serde.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use serde;
+use serde::de::{Deserialize, Deserializer, Error};
+use serde::ser::Serializer;
+use serde::Serialize;
+use serde_with::{Bytes, DeserializeAs, SerializeAs};
+
+/// Encode bytes to Hex for human readable serializer and deserializer,
+/// serde to Bytes for non human readable serializer and deserializer.
+pub trait BytesOrHex<'de>: Sized {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+        Self: AsRef<[u8]>,
+    {
+        serialize_to_bytes_or_encode(serializer, self.as_ref(), Encoding::Hex)
+    }
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>;
+}
+
+/// Encode bytes to Base64 for human readable serializer and deserializer,
+/// serde to array tuple for non human readable serializer and deserializer.
+pub trait BytesOrBase64<'de>: Sized {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+        Self: AsRef<[u8]>,
+    {
+        serialize_to_bytes_or_encode(serializer, self.as_ref(), Encoding::Base64)
+    }
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>;
+}
+
+fn deserialize_from_bytes_or_decode<'de, D>(
+    deserializer: D,
+    encoding: Encoding,
+) -> Result<Vec<u8>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    if deserializer.is_human_readable() {
+        let s = String::deserialize(deserializer)?;
+        encoding
+            .decode(s)
+            .map_err(|_| D::Error::custom("byte deserialization failed"))
```

Comment:
Whats the reason for throwing away the error from decoding, wouldn't it be useful to include that in the error message  (via the error's `Display` impl) so that a user can get a better idea of why the deserialization failed? Eg from this error I don't know why it failed, did i give it hex when it wanted base64 or vice versa

Reply:
👍thanks for pointing this out, I will add more info to the error